src/Datadog.Trace/Immutables/DatadogImmutableStack.cs

```diff
@@ -0,0 +1,234 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Datadog.Trace.Immutables
+{
+    internal sealed class DatadogImmutableStack<T> : IDatadogImmutableStack<T>
```

Comment:
Note: Including the `System.Collections.Immutable` assembly caused some versioning issues in netstandard, will post more details when I get back to this.
The fix, which I think is reasonable to reduce our dependencies, is to have a clone of the immutable stack. It's not the most complex piece of code ever, it stores a reference to the last stack with a new instance and keeps the current level as `_head`. To pop, you return the previous instance, to push you return a new instance. It removes the risk of overwriting other call contexts when pushing or popping. This was inspired by the discovery that beginning an AsyncLocal early enough in the dot net http pipeline means the call context is actually shared across all of the concurrent http contexts (not ideal).